grails-app/services/au/org/ala/biocache/hubs/WebServicesService.groovy

```diff
@@ -659,6 +659,32 @@ class WebServicesService {
         dataQualityCodes
     }
 
+    // maps from country name to iso code
+    @Cacheable('longTermCache')
+    def getCountryNameMap() {
+        def countryUrl = "${grailsApplication.config.userdetails.baseUrl}/ws/registration/countries.json"
+        def countries = getJsonElements(countryUrl)
+        return countries?.findAll {it -> beAValidCountryOrState(it as JSONObject)}?.collectEntries { [(String)it.get("name"), (String)it.get("isoCode")] }
+    }
+
+
+    private static boolean beAValidCountryOrState(JSONObject obj) {
+        return obj.has("isoCode") && obj.has("name") && obj.get("isoCode") != "" && obj.get("name") != "N/A"
+    }
+
+    @Cacheable('longTermCache')
+    List<String> getStates(String countryName) {
+        Map countryNameMap = grailsApplication.mainContext.getBean('webServicesService').getCountryNameMap()
+        // if a known country name
+        if (countryNameMap?.containsKey(countryName)) {
+            def states = getJsonElements("${grailsApplication.config.userdetails.baseUrl}/ws/registration/states.json?country=" + countryNameMap.get(countryName))
```

Comment:
Just checking if this needs to be (exception) caught? What happens is the service throws an error? Does the user (or ALA staff) have anyway to know that its not working or will it silently fail and be an empty list (that you could interpret as "we don't have any states for this country"?

Reply:
Hi Nick, for frontend, only when request is successful will be state list be re-constructed. 
`$.getJSON(url, function(data) `

1. If not empty, populate it
2. If empty (no matter because originally empty or backend error), populate with the original list.

Reply:
For the backend.

```
def getCountryNameMap() {
    def countryUrl = "${grailsApplication.config.userdetails.baseUrl}/ws/registration/countries.json"
    def countries = getJsonElements(countryUrl)
```

`getJsonElements` might throw an exception, why it doesn't have a throws in definition 
`JSONElement getJsonElements(String url, String apiKey = null) {` 

I should catch it I think